### PR TITLE
Fixes #140392 - Implement workbench.action.exitZenMode

### LINKED ISSUE #140392: Command "workbench.action.exitZenMode" would turn on ZenMode when not in ZenMode
Issue Type: <b>Bug</b>

1. Add a shortcut like
  ```json
    {
        "key": "alt+e alt+w",
        "command": "workbench.action.exitZenMode",
    },
  ```
2. ensure not in ZenMode
3. fire the shortcut "alt+e alt+w"
4. the ZenMode would be turned on

i.e. `exitZenMode` behaves just like `toggleZenMode` ...

---

Does this issue occur when all extensions are disabled?: Yes

VS Code version: Code 1.63.2 (899d46d82c4c95423fb7e10e68eba52050e30ba3, 2021-12-15T09:40:02.816Z)
OS version: Windows_NT x64 10.0.19043
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-8265U CPU @ 1.60GHz (8 x 1800)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: enabled_on<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|7.80GB (0.73GB free)|
|Process Argv|--crash-reporter-id aef19635-1b79-4f09-9c0d-1ff29d49308d|
|Screen Reader|no|
|VM|50%|
</details><details><summary>Extensions (17)</summary>

Extension|Author (truncated)|Version
---|---|---
dendron|den|0.76.0
dendron-paste-image|den|1.1.0
githistory|don|0.6.19
todo-tree|Gru|0.0.214
vscode-wordcount-cjk|hol|1.3.1
vscode-settings-cycler|hoo|1.0.1
path-autocomplete|ion|1.18.0
python|ms-|2021.12.1559732655
wordcount|ms-|0.1.0
markdown-checkbox|PKi|1.7.2
vscode-yaml|red|1.2.2
multi-command|ryu|1.5.1
cjk-word-handler|Sha|0.1.1
pdf|tom|1.2.0
markdown-all-in-one|yzh|3.4.0
material-theme|zhu|3.13.10
writing-extension-pack-for-binly|N/A|0.0.1

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
pythontb:30283811
pythonptprofiler:30281270
vsdfh931cf:30280410
vshan820:30294714
vstes263cf:30335440
vscoreces:30384385
pythondataviewer:30285071
vscod805:30301674
pythonvspyt200:30340761
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
vsaa593cf:30376535
pythonvs932:30410667
vscop804:30404766
vscop940:30404999
vsrem710cf:30416617
py55gd98cf:30419166
vscexrecpromp3t1:30407762

```

</details>

---

usage scenario:
Since the config setting `zenMode.hideTabs` (after toggled) would not be effective until re-enter the ZenMode (if already in ZenMode), and I only need `hideTabs==true` for few scenario, the command `exitZenMode` would help for command macros.

workaround:
define multiple shortcuts depending on the when clause "inZenMode" and "!inZenMode", which may be very troublesome when combined with another when clauses ... (especially without paranthesis... https://github.com/microsoft/vscode/issues/91473

### PR DESCRIPTION
**Pull request by ghost**
This PR fixes #140392

- [x] I've gone over this issue on my own

## Issue Summary

According to the original submitter:

> Command "workbench.action.exitZenMode" would turn on Zen Mode when not in ZenMode

This occurred because `exitZenMode` would call `toggleZenMode`, which fires regardless of whether Zen Mode is on.

## Proposed Changes

Added `isInZenMode`. Switched `exitZenMode` to call `toggleZenMode` only when `isInZenMode` reports the editor is in Zen mode instead.

## Testing

1. Switch to the build of the feature branch.
2. Assign a hotkey to `workbench.action.exitZenMode`.
    - suggested defaults that don't collide with other shortcuts are `alt+e alt+w` for Windows or Linux, or if on Mac, `alt+cmd+g`
3. Try using the hotkey when not in Zen Mode. Ensure that it does not toggle Zen Mode on.
4. Try using the hotkey when in Zen Mode (on Mac, Zen Mode can be toggled on by default via `cmd-k z`). Ensure that it toggles Zen Mode off.

### CODE REVIEW

File: src/vs/workbench/browser/actions/layoutActions.ts [hunk 1]
```diff
@@ -463,7 +463,7 @@ KeybindingsRegistry.registerCommandAndKeybindingRule({
 	weight: KeybindingWeight.EditorContrib - 1000,
 	handler(accessor: ServicesAccessor) {
 		const layoutService = accessor.get(IWorkbenchLayoutService);
-		layoutService.toggleZenMode();
+		layoutService.toggleZenModeOff();
```
Comment (JacksonKearl) [inline src/vs/workbench/browser/actions/layoutActions.ts:466]: I'd rather we query the value of the `InEditorZenModeContext` context key dynamically here and only dispatch to `toggleZenMode` when it's true, adding a whole separate method to the core layout service to half-way duplicate an existing method isn't ideal. Alternatively we could adjust `toggleZenMode` to take the desired end state as an optional parameter, similar to how `DOMNode#classList.toggle` operates.

**Review by JacksonKearl — CHANGES_REQUESTED** [review]

File: src/vs/workbench/browser/actions/layoutActions.ts (hunk 1 above)
Comment (chenjefferson) [reply src/vs/workbench/browser/actions/layoutActions.ts]: You're right, and I learned some architecture today. Thanks!

I guess the issue is `InEditorZenModeContext` is stored in the layout service's private state, and its public interface doesn't support parameters to `toggleZenMode` as some args change internal state. New set of changes add `isInZenMode` to expose that and use that instead, but let me know if there's a better solution!

Comment (JacksonKearl) [reply src/vs/workbench/browser/actions/layoutActions.ts]: > I guess the issue is InEditorZenModeContext is stored in the layout service's private state

In fact all context keys are almost always global*, but they need to be accessed via an `IContextKeyService`. You can see an example of how this looks here: https://github.com/microsoft/vscode/blob/2cf8ae8a1394e7428c840000a48c621f986a4660/src/vs/workbench/contrib/debug/browser/debugCommands.ts#L297-L302

*some context keys are "scoped", and only apply to children of specific HTML elements. This isn't the case here.

You can see the full list of context keys applied to a given HTML element via the `Developer: inspect Context Keys` command. (result is logged to the dev tools)

**Review by JacksonKearl — CHANGES_REQUESTED** [review]
Use dynamic context querying as discussed above (just submitting a real review to help with my inbox tracking)

**Comment by chenjefferson** [comment]
No worries! Currently swamped with work. Will ping when I get a chance to implement this

**Comment by JacksonKearl** [comment]
Thanks for working on this! I will be closing it out in favor of #144217

**Pull request closed without merging**